lib/pengine/utils.c

```diff
@@ -1263,9 +1264,28 @@ unpack_operation(pe_action_t * action, xmlNode * xml_obj, pe_resource_t * contai
         }
     }
 
-    value = g_hash_table_lookup(action->meta, XML_ATTR_TIMEOUT);
+    value = NULL;
+    if (pcmk__str_any_of(action->task, RSC_START, RSC_STATUS, NULL)
+        && action->rsc->parameters && action->rsc->xml
+        && safe_str_eq(crm_element_value(action->rsc->xml,
+                                         XML_AGENT_ATTR_CLASS),
+                       PCMK_RESOURCE_CLASS_STONITH)) {
+
+        const char *st_mon_timeout;
+        st_mon_timeout = g_hash_table_lookup(action->rsc->parameters,
+                                             "pcmk_monitor_timeout");
+        if (st_mon_timeout) {
+            value = st_mon_timeout;
```

Comment:
I'm thinking we should default to pcmk_monitor_timeout * pcmk_monitor_retries (i.e. only if there is no explicitly specified meta timeout).

Reply:
> I'm thinking we should default to pcmk_monitor_timeout * pcmk_monitor_retries 

Sounds reasonable.

> (i.e. only if there is no explicitly specified meta timeout).

So, explicitly specified meta timeout overrides (pcmk_monitor_timeout * retries), which overrides default meta timeout. I think the final override with the explicit meta timeout is another departure from the current behavior, so I'll want to check what effect that has. I can understand giving an explicit meta timeout the final say, from a UX perspective. Though we may want to go with max(explicit_meta, (pcmk_monitor_timeout * retries)) as you suggested earlier.

I'll also need to decide how best to flag an explicit meta timeout versus a default one. For example, if a user sets an explicit meta timeout of 20s, which matches the default.

Reply:
> I'll also need to decide how best to flag an explicit meta timeout versus a default one. For example, if a user sets an explicit meta timeout of 20s, which matches the default.

Timeouts are more complicated than the average default, but it's still straightforward earlier in the function. If you see "Handle timeout default", the lookup there is exactly whether the user specified an explicit value (either in the op element directly or in a meta-attributes block). The else block is where the default gets injected if needed.

Reply:
Ack thanks. I responded that last time without going back and looking at unpack_operations again, sorry. I assumed it might be more complicated than that.

Reply:
> I'm thinking we should default to pcmk_monitor_timeout * pcmk_monitor_retries

pcmk_monitor_retries only applies WITHIN pcmk_monitor_timeout. "pcmk_monitor_timeout=30s and pcmk_monitor_retries=2" doesn't allow us to go up to 60 seconds. The action times out after 30 seconds no matter what. On the other hand, if we fail after 20 seconds, then we get another try in the 10 seconds that remain.

So setting the op meta timeout to (pcmk_monitor_timeout * retries) doesn't seem to make any difference.

Reply:
> (i.e. only if there is no explicitly specified meta timeout).

After testing with "explicit meta timeout overrides pcmk_monitor_timeout", this leaves us open to continued weird-to-the-end-user behavior IMO.

Allowing an explicit meta timeout to take precedence concerns me for a couple of reasons.

1. The op meta timeout effectively **is not** enforced for the recurring monitor operation, but it **is** enforced for the start operation. So we still end up with a confusing behavioral difference at the end user level.
    - As you explained, the recurring monitor gets scheduled and returns immediately. The stonith monitor command within the fencer is thus allowed the full pcmk_monitor_timeout even if the op meta timeout is set to a lower value (e.g., by unpack_operation).
    - The start operation waits for execd_stonith_monitor to finish. The stonith monitor command is allowed the full pcmk_monitor_timeout, but the start operation that called it is not.

2. If the explicit meta timeout for the start op is less than pcmk_monitor_timeout, we can still end up with a premature start timeout while the monitor command is running, and get an "action lost" message. That's the behavior that we're trying to prevent. In this case it would be the user's fault for setting a lower meta timeout for the start operation. But I have trouble seeing a use case where making the start time out early (before monitor command completes) is the desired behavior.

    - Even though it would time out early, it would still wait for the cluster-delay to expire waiting for a result, causing a further delay to recovery.

It seems to me that the following order of precedence for stonith start/monitor ops is the most practical:
1. pcmk_monitor_timeout
2. explicit meta timeout
3. op default timeout
4. CRM_DEFAULT_OP_TIMEOUT_S

Let me know if you think differently. Deciding how this **ought** to behave has been tough.

Reply:
That seems reasonable (assuming #2 means explicit timeout whether in op tag or meta attributes)